cogs/staffapps_frontend.py

```diff
@@ -104,18 +106,20 @@ def ongoing(self) -> discord.Embed:
         ).build()
 
 
-class user:
-    def __init__(self, user: discord.Member) -> None:
+class Applicant:
+    def __init__(self, user: discord.User | discord.Member) -> None:
         """
         This class represents a read-only user who is applying for staff.
         Data is accurate at the time of object creation.
         """
+        self.user = user
+
         # connect to database
         db = database.connect()
         cursor = db.cursor()
         # get user data from discord
         self.uid = user.id
-        self.joined_at = user.joined_at
+        self.joined_at = getattr(user, "joined_at") or datetime.now()
```

Comment:
See comment from Czar, discord.User will always have a joined_at attr.
Also using datetime.now() is impossible (A user MUST be in the server for >30 days to apply, how is it possible they joined now?)

Reply:
`discord.Member` will always have a `joined_at`, but `discord.User` will never have one (since a User would come from a DM).

This class accepts `user: discord.User | discord.Member` to simplify the typing of the class instantiation, so the typing system doesn't know whether or not `user` will have a `joined_at`. 

The class accepts `discord.User | discord.Member` (rather than just `discord.Member`) because unfortunately the `author` from `discord.ApplicationContext` is typed that way; pycord will not allow us to narrow down that type such that the slash command must have been used by a server member.

The point of using `datetime.now()` is specifically because it's impossible - the user is guaranteed to fail the prequalification checks if they are missing a `joined_at` (i.e. if they used the slash command in DMs).

Let me shuffle around the typing to make this a little more straightforward... Is this better? f68ec98ddb6f1033bde37d3d5b0deafd5cf497a3

Reply:
Ohh, yeah I get what you were trying to do now.. I like the way you resolved this in https://github.com/Arborym/Pax-Academia/commit/b4e0795d2a2e927658c82e707adb14945045d0d9 .